### ISSUE TITLE
Unoptimized IR generates `linkersymbol("...")` for internal library calls

### ISSUE THREAD

**Issue by hrkrshnn** [ethereum/solidity#12865]
```solidity
library Math {
    function add(uint a, uint b) internal returns (uint c) {
        c = a + b;
    }
}
contract C {
    function f(uint a, uint b) public returns (uint c) {
        c = Math.add(a, b);
    }
}
```

The IR for the above generates `let expr_27_address := linkersymbol("<stdin>:Math")`. The optimizer is able to optimize it away. However, the compiler shouldn't any code with `linkersymbol` when it's not needed, even in the unoptimized codegen.

**Comment by cameel** [comment]
Good point. I agree that we should avoid leaving these linker references in when the call is actually internal, even in unoptimized code. We don't want people to have to go through some dummy linking process when these references are unused.

This is not actually a bug though, it was done like this on purpose. At least there was no consideration of removing it at the codegen level when this was being implemented. The evaluation of the name into address and actual use of the address happen at different points in AST processing and this follows our general pattern of just creating an IR variable without checking if it's used or not and then using it (or not) in a another place, assuming that the optimizer will remove it if it's unused.

The way it's done currently, the codegen simply puts the library address in a variable whenever it encounters its name:
https://github.com/ethereum/solidity/blob/c4909e99c1015233d6937d229058cde52615f246/libsolidity/codegen/ir/IRGeneratorForStatements.cpp#L2138-L2142

If it happens to be an external call, then `appendExternalFunctionCall()` called when visiting a `FunctionCall` may then use that variable:
https://github.com/ethereum/solidity/blob/c4909e99c1015233d6937d229058cde52615f246/libsolidity/codegen/ir/IRGeneratorForStatements.cpp#L2580-L2581

Looks like we'll have to complicate this mechanism a bit. Similar to the way that having to detect actual calls (rather than all uses of a function name) complicates collecting functions for internal dispatch.

**Comment by chriseth** [comment]
It was done like this intentionally. We should not complicated code generation if a simple optimizer step can do the same trick. If people don't want to cope with the longer compilation time of the optimizer, we can offer a minimal optimizer sequence that takes care of it.

**Comment by hrkrshnn** [comment]
We may have to fix this in the codegen. Leaving `--yul-optimizations=""` is currently used as a proxy for "debug" mode for faster build times. This just adds linking, which is unnecessary for 99% of codebases.

**Comment by cameel** [comment]
It's a bit easier to do this at the optimizer level though. In the codegen the way we do this, is basically to insert `linkersymbol()` whenever we see a library name, without checking if it's necessary or not - and we don't know that until later where we actually visit an identifier or member access node. Maybe it would be easier to have a limited version of the step that removes unused variables just for this?

**Comment by chriseth** [comment]
Yes, please let us not do conditional codegen again. It is a nightmare in the old code generator. Let's just run simple unused variable remover or something, or just recommend that to the user in some way.

**Comment by ekpyron** [comment]
We should make some basic optimizations (including stack-to-memory) the default for via-IR and be done with this that way. Has been requested by tooling anyways.

**Comment by cameel** [comment]
This is was probably solved by #14149, which will be included in the next release, just need to confirm.

**Comment by nikola-matic** [comment]
Closing this as solved since https://github.com/ethereum/solidity/pull/14149 was merged, which enables (amongst other things) the UnusedPruner optimization step (`u`) by default. You can confirm via `solc file.sol --ir-optimized`.
